### PR TITLE
Deferred call to .OnApplyTemplate()

### PR DESCRIPTION
**Pull request by carldebilly**
# Bugfix
Following discoveries in #3519, this PR is adressing the deferred call to `.ApplyTemplate()` on all platforms.

## What is the new behavior?
* `.OnApplyTemplate` is now called right after the `Loading` event.
* The behavior is by default on Wasm. Can be turned off by putting this code in the startup of the app:
  ``` csharp
   FeatureConfiguration.Control.UseDeferredOnApplyTemplate = false;
  ```
* Opt-in on other platforms. Turn it on with this code in the startup of the app:
  ``` csharp
   FeatureConfiguration.Control.UseDeferredOnApplyTemplate = true;
  ```
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [X] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [X] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [X] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/Uno.UI.Tests/Windows_UI_Xaml/Given_StaticResource.cs [hunk 1]
```diff
@@ -147,6 +147,8 @@ public void When_Inherited_In_Template_Applied()
 			control.InlineTemplateControl.ApplyTemplate();
 			control.TemplateFromResourceControl.ApplyTemplate();
 
+			control.ForceLoaded();
```
Comment (davidjohnoliver) [inline src/Uno.UI.Tests/Windows_UI_Xaml/Given_StaticResource.cs:150]: Is this necessary? Shouldn't ApplyTemplate() be sufficient?

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml/UIElementTests/UIElement_InitializationSequence_TestControl.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		{
+			_logger = logger;
+			Name = name;
+			using (LogRecursively("ctor"))
+			{
+				DefaultStyleKey = typeof(UIElement_InitializationSequence_TestControl);
+
+				Loaded += (snd, evt) => Log("Loaded");
+				Loading += (snd, evt) => Log("Loading");
+				//LayoutUpdated += (snd, evt) => Log("LayoutUpdated");
```
[the remark this answers is not in the document]
Comment (davidjohnoliver) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml/UIElementTests/UIElement_InitializationSequence_TestControl.cs:21]: +1, or add a comment as to why it's commented ;)

File: src/Uno.UI.Tests/Windows_UI_Xaml/Given_ThemeResource.cs [hunk 3]
```diff
@@ -85,6 +87,8 @@ public void When_Inherited_In_Template_Applied()
 			control.InlineTemplateControl.ApplyTemplate();
 			control.TemplateFromResourceControl.ApplyTemplate();
 
+			control.ApplyTemplate();
```
Comment (davidjohnoliver) [inline src/Uno.UI.Tests/Windows_UI_Xaml/Given_ThemeResource.cs:90]: Necessary? `Test_Control` doesn't have a template, it's a `UserControl`.

File: src/Uno.UI/UI/Xaml/Controls/Control/Control.cs [hunk 4]
```diff
@@ -164,12 +165,38 @@ internal View TemplatedRoot
 					{
 						RegisterContentTemplateRoot();
 
-						OnApplyTemplate();
+						if (FeatureConfiguration.Control.UseDeferredOnApplyTemplate)
+						{
+							// It's too soon the call the .ApplyTemplate method: it should be invoked right before the Loading event.
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/Control/Control.cs:170]:
```suggestion
							// It's too soon to call the .ApplyTemplate method: it should be invoked after the Loading event.
```

right?

File: src/Uno.UI/UI/Xaml/FrameworkElement.wasm.cs [hunk 5]
```diff
@@ -26,6 +26,10 @@ public partial class FrameworkElement : IEnumerable
 
 		partial void OnLoadingPartial();
 
+		internal virtual void OnPostLoading()
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/FrameworkElement.wasm.cs:29]: should this be `private protected`? Also why not just declare it once in the shared `FrameworkElement.cs` file?

File: src/Uno.UI.Tests/Windows_UI_Xaml/Given_StaticResource.cs (hunk 1 above)
Comment (carldebilly) [reply src/Uno.UI.Tests/Windows_UI_Xaml/Given_StaticResource.cs:150]: It was required in this case. I didn't investigate why exactly.

File: src/Uno.UI.Tests/Windows_UI_Xaml/Given_ThemeResource.cs (hunk 3 above)
Comment (carldebilly) [reply src/Uno.UI.Tests/Windows_UI_Xaml/Given_ThemeResource.cs:90]: Yep it is. The control was failing without this.

File: src/Uno.UI/UI/Xaml/Controls/Control/Control.cs (hunk 4 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Control/Control.cs:170]: Actually it's the call to ".OnApplyTemplate".  I'll fix it.  Nice catch!

File: src/Uno.UI/UI/Xaml/Controls/Control/Control.cs [hunk 6]
```diff
@@ -164,12 +165,38 @@ internal View TemplatedRoot
...
+			base.OnPostLoading();
+
+			TryCallOnApplyTemplate();
+		}
+
+		private void TryCallOnApplyTemplate()
+		{
+			if (_applyTemplateShouldBeInvoked)
+			{
+				_applyTemplateShouldBeInvoked = false;
```
Comment (dr1rrb) [inline src/Uno.UI/UI/Xaml/Controls/Control/Control.cs:195]: If in the handler of the loading does `control.ApplyTemplate()`, the `OnApplyTemplate` will be invoked twice. You should reset the flag also in the `ApplyTemplate()`

**Pull request merged**